paasta_tools/mesos_tools.py

```diff
@@ -678,9 +678,13 @@ def get_mesos_task_count_by_slave(mesos_state, slaves_list=None, pool=None):
     :returns: list of slave dicts {'task_count': SlaveTaskCount}
     """
     all_mesos_tasks = get_all_running_tasks()  # empty string = all app ids
-    slaves = {
-        slave['id']: {'count': 0, 'slave': slave, 'chronos_count': 0} for slave in mesos_state.get('slaves', [])
-    }
+    try:
+        slaves = {
+            slave['id']: {'count': 0, 'slave': slave, 'chronos_count': 0} for slave in mesos_state.get('slaves', [])
+        }
+    except SlaveDoesNotExist:
+        log.debug("Tried to get mesos slaves for id {}, but none existed.".format(mesos_state.get('id'), ''))
+        return {}
     for task in all_mesos_tasks:
         if task.slave['id'] not in slaves:
```

Comment:
@jglukasik I think this is the line that is blowing up.

So I think you could do something like:
```
try:
  if task.slave['id'] not in slaves:
    continue
except SlaveDoesNotExist:
  continue
```
although that exact formation is a bit nasty to read... This whole function is a bit of a mess (and I think I am probably to blame)!

The other thing I would suggest is trying to TDD this. I haven't checked what the unit test looks like but if it's easy to add a sideeffect to the right mock then you should be able to check that you've fixed it more easily.